### ISSUE TITLE
marvellous

### ISSUE THREAD

**Issue by stef-ladefense** [RobTillaart/map2colour#6]
thank you very much for your library
it works perfectly
it allows me to make dials with a fade of several colors

**Comment by RobTillaart** [comment]
Thank you, good to hear it works as intended (that is an assumption)
Any chance you have a video/youtube link?

**Comment by stef-ladefense** [comment]
![image](https://user-images.githubusercontent.com/43587190/196563119-0458d417-bfee-44a2-90f0-293a949f7c30.png)
left: map2colour
right: map2colourFast

but I don't know if it's normal to have only non-gradient colors on the right

**Comment by RobTillaart** [comment]
> but I don't know if it's normal to have only non-gradient colors on the right

mmm should not happen, need to dive into the details of the code. Might take a few days.

**Comment by RobTillaart** [comment]
@stef-ladefense

Checked code today and the output of map2colour and map2colourFast should be identical.
Added a example that does compare mapping and on my UNO these were identical.
Added this test to the unit test too.
Created a develop branch with the above and some more (minor) changes.

- What board are you using?
- Can you create a minimal sketch that shows the difference of the math of the gradient and non gradient arcs?

**Comment by stef-ladefense** [comment]
I will make a mini code for it
I am using an ESP32S2 with arduino-esp32 v2.05

**Comment by RobTillaart** [comment]
Super, will have a look tomorrow (22:00 here)

**Comment by stef-ladefense** [comment]
```cpp
// 240*240 TFT 8-pins SPI interface GC9A01 controller
// NOTE the display is a 3.3V device.
// uses TFT _eSPI library with modified setup.h file (in the library folder) or use build_opt.h (only for esp32)
// ESP32S2    TFT_MOSI:11    TFT_SCLK:7   TFT_CS:12    TFT_DC:13    TFT_RST:10    TFT_BL:14

#include <SPI.h>
#include <TFT_eSPI.h>     // https://github.com/Bodmer/TFT_eSPI
TFT_eSPI tft = TFT_eSPI();

#include "map2colour.h"   // https://github.com/RobTillaart/map2colour
map2colour mc;
//map2colourFast mc;
float values[7] = { -200, -90, 0, 45, 150, 180, 180 };
uint32_t colours[7] = { M2C_BLUE, M2C_AQUA, M2C_LIME, M2C_YELLOW, M2C_RED, M2C_RED, M2C_RED };

void setup() {
  mc.begin(values, colours);

  tft.init();
  tft.setRotation(0);  // 0, 1, 2 ou 3
  tft.fillScreen(TFT_BLACK);

  for (int16_t i = -150; i <= 150; i += 1) {
    float _precalc = (i - 90) * 0.0174532925;
    float sx = cos(_precalc) * 108;
    float sy = sin(_precalc) * 108;
    tft.fillCircle(120 + sx, 120 + sy, 11, mc.map2_565(i));
  }
}

void loop() {
}
```

**Comment by stef-ladefense** [comment]
with map2colour mc; this work
with map2colourFast mc; dont

but with float values[7] = { -200, -90, 0, 45, 150, 180, **181** }; work too

**Comment by RobTillaart** [comment]
Can you check the return value of **mc.begin(,,,)**?
Should return true == 1

**Comment by RobTillaart** [comment]
Check the readme.md
The value array should be ascending.
As this is not the case the begin() of the fast class will not calculate the divfactors.
This prevents interpolation thus no gradient.

**Comment by stef-ladefense** [comment]
yes I saw after, now it works very well with the modification so that all the values are different.
I let myself be fooled since it still worked with the non-fast version when several values are identical

**Comment by stef-ladefense** [comment]
Thanks again for sharing your work.
it's great to be able to adjust the spectrum so finely

**Comment by RobTillaart** [comment]
Good to hear it is solved, however your issue gave me an idea to allow identical numbers, that would allow to make gradients with a break in it.

Imagine these values.
```cpp
float values[7] = { -200, -90, 0, 45,                                   45, 150, 180 };
uint32_t colours[7] = { M2C_BLUE, M2C_AQUA, M2C_LIME, M2C_YELLOW,       M2C_RED, M2C_YELLOW, M2C_BLUE};
```

with the double 45 in the values there would be no gradient between the **M2C_YELLOW** and **M2C_RED**
effectively having 2 adjacent gradients.

I'm going to add this into the develop branch. 


----

With your code I do not know how often you draw the gradient but I think you can speed it up,
assuming there is a line function in the tft lib. Only risk is that it would not do a complete fill.

```cpp
  for (int16_t i = -150; i <= 150; i += 1) {
    float _precalc = (i - 90) * 0.0174532925;
    float sx = cos(_precalc) * 108;
    float sy = sin(_precalc) * 108;
    //  tft.fillCircle(120 + sx, 120 + sy, 11, mc.map2_565(i));
   tft.line(120 + x - 5, 120 + y - 5, 120 + x + 5, 120 + y + 5, mc.map2_565(i));
  }
```

**Comment by stef-ladefense** [comment]
thank you very much, I had started similar but I had unfilled areas, after I found an example in tft with triangles on 2 coordinates of circles which was good, then I wanted the start, end and middle in circle, and it was the simplest, as the circle only updates every 5 seconds, it doesn't matter.

the code I sent you has been simplified for testing.
in reality the construction is slightly different, it starts the construction by the maximun towards the minimum from right to left.
which allows to have a gray background color, then the spectrum, all with circles at the beginning, at the end of the spectrum and also at the end of the gray

```cpp
void CadranValeur(uint16_t value) {

  const uint16_t _vmin = 400;
  const uint16_t _vmax = 1000;
  const uint8_t _diam = 11;
  const uint16_t _rayon = _CIRCLE_Centre_X_ - _diam - 1;
  const uint8_t _angle = 150;

  int16_t _V_angle = map(constrain(value, _vmin, _vmax), _vmin, _vmax, -_angle, _angle);

  for (int16_t i = _angle; i >= -_angle; i -= 1) {
    float _precalc = (i - 90) * 0.0174532925;
    float sx = cos(_precalc) * _rayon;
    float sy = sin(_precalc) * _rayon;
    uint16_t _C_ = i <= _V_angle ? (value <= _vmax ? mc.map2_565(i) : TFT_RED) : _ColorTFT_Fond_Jauge;
    tft.fillCircle(_CIRCLE_Centre_X_ + sx, _CIRCLE_Centre_Y_ + sy, _diam, _C_);
  }
}
```

**Comment by RobTillaart** [comment]
@stef-ladefense 
Released 0.1.5 which solves your initial issue.
If there are no other related questions, please close the issue

ps if you post code add cpp after backquotes to get syntax 
```
//  ```cpp
```

**Comment by stef-ladefense** [comment]
I just tested the 0.15, it works without worries.
thanks again
